Search and filter of Garden Lists

MulesKC opened the pull request:
… selecting a feature during garden updates

finished search feature and category filter feature; fixed bug heather found w/ viewing owners w/o any gardens in db

fixing bugs with routing

fixed titling of various sites, got /admin to disappear from navbar when active site, removed search.html - not used

added view gardens by feature to index.html; updated HomeController and GardenController as needed for this addition

added /user/ to several links in various templates

MulesKC committed:
updated gardens/list and view to Heather's latest; fixed bug with not selecting a feature during garden updates

finished search feature and category filter feature; fixed bug heather found w/ viewing owners w/o any gardens in db

fixing bugs with routing

fixed titling of various sites, got /admin to disappear from navbar when active site, removed search.html - not used

added view gardens by feature to index.html; updated HomeController and GardenController as needed for this addition

added /user/ to several links in various templates

small html updates; removed several else statements from GardenData

fixed merge conflicts in index and gardens/add

scottpenneman commented on src/main/resources/templates/gardens/list.html:
```diff
@@ -12,8 +12,50 @@
+            </div>
+        </form>
+    </div>
+</div>
```
This div closing tag can be deleted, as it prematurely closes the div tag on line 14. Without line 46, the tag is still closed on line 74.

MulesKC replied:
Good catch! Got it updated.

scottpenneman reviewed (approved):
I left one comment regarding a div tag, but otherwise it worked just fine for me! Awesome feature for the page as well!

KBeas21 commented on src/main/java/org/launchcode/water_garden_tour/controllers/HomeController.java:
```diff
@@ -1,20 +1,24 @@
+    @Autowired
+    private FeatureRepository featureRepository;
+
+    @GetMapping("")
```
why did we change RequestMapping to GetMapping here?

MulesKC replied:
I thought it was the more correct annotation since it specifies the type of request. It was working fine before so I can definitely leave it as RequestMapping.

KBeas21 replied:
sounds good to me thanks for the explanation!

KBeas21 commented on src/main/resources/templates/gardens/list.html:
```diff
@@ -12,8 +12,50 @@
+                </th:block>
+                <p>Note: all gardens are searched if no categories are selected.</p>
+            </div>
+            <div>
```
are the empty div tags here and closing it on 43 needed for anything? If they are needed, what are they doing?

MulesKC replied:
I thought I was using them around the button but I removed them and nothing changed so yeah they were definitely unneeded.

KBeas21 commented on src/main/java/org/launchcode/water_garden_tour/models/garden/GardenData.java:
```diff
@@ -0,0 +1,102 @@
+            for (Integer featureId : featureIds) {
+                curFeature = getFeatureFromId(featureId, allFeatures);
+
+                if (garden.getFeatures().contains(curFeature)) {
+                } else {
+                    hasAllFeatures = false;
+                }
```
you could simply not what is inside of the if statement and then do away with the else case here:
`if (garden.getFeatures().contains(curFeature)) { hasAllFeatures = false } `

MulesKC replied:
Good call! Update made.

KBeas21 commented on src/main/java/org/launchcode/water_garden_tour/models/garden/GardenData.java:
```diff
@@ -0,0 +1,102 @@
+            }
+        }
+
+        if (searchValue.equals("all") || searchValue.equals("")){
+
+        } else {
+            List<Garden> iteratingGardenList = new ArrayList<>();
+            iteratingGardenList.addAll(results);
+
+            for (Garden garden : iteratingGardenList) {
+                if (!isTextInGardenFields(garden, searchValue)) {
+                    results.remove(garden);
+                }
+            }
+
+        }
```
inverse this logic like in the above comment and remove the else

KBeas21 commented on src/main/java/org/launchcode/water_garden_tour/models/garden/GardenData.java:
```diff
@@ -0,0 +1,102 @@
+
+        if (searchValue.equals("all") || searchValue.equals("")){
+            results = allGardens;
+        } else {
+
+        }
```
no need for else

KBeas21 commented on src/main/java/org/launchcode/water_garden_tour/models/garden/GardenData.java:
```diff
@@ -0,0 +1,102 @@
+                return feature;
+            }
+        }
+        return null;
```
no reason to do anything here but be careful when returning null values as when you call this in the future you might get null pointers and think "Hey my objects are empty or not holding anything"

MulesKC replied:
That makes sense, duly noted

KBeas21 commented on src/main/java/org/launchcode/water_garden_tour/models/garden/GardenData.java:
```diff
@@ -0,0 +1,102 @@
+
+    public static boolean isTextInGardenFields(Garden curGarden, String searchValue) {
+
+            if (curGarden.getName().toLowerCase().contains(searchValue) ||
+                    curGarden.getDescription().toLowerCase().contains(searchValue) ||
+                    curGarden.getAddress().toLowerCase().contains(searchValue) ||
+                    curGarden.getOwner().getName().toLowerCase().contains(searchValue)) {
```
because this call is doing a lot I might create a method in the Garden class to do this and then call that here. Just an idea you don't have to

MulesKC replied:
I was using this chunk of code a few times in my search methods above so I decided to make it it's own method. I put it here in GardenData because it gets used here but could put it in Garden if that is more appropriate.

KBeas21 replied:
nope here is fine just wanted to know more about it good work!

jamiander reviewed (commented):
I really like this!  
One thing, Admin doesn't disappear from the nav bar.  I think you need to add "admin-link "to the fragments.html.

KBeas21 reviewed (approved):
good work +1

hstrodtman commented:
Awesome job!

MulesKC committed:
removed controller methods from GardenController that are now in AdminController

MulesKC committed:
fixed double welcome picture in index

MulesKC commented:
I removed the redundant methods from GardenController, and only had updates for AdminController where my branch had changes. Found a small bug in index that I removed. Everything was straightforward and looks to be working.
